Provide modules with information how much a quantity changed

Hlavtox opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | This PR allows modules to know how much the quantity has changed during update. Not only the new number.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Partially fixes #25981
| How to test?      | Test using ps_qualityassurance and see that correct increments/decrements are provided in delta_quantity
| Possible impacts? | -

I named the parameter the same as in other parts of PS: https://github.com/PrestaShop/PrestaShop/blob/ad181f7fb89dea67dfebba779a6440854baa6d5b/src/Core/Stock/StockManager.php#L193

Hlavtox committed:
Provide modules with information how much a quantity changed

khouloudbelguith commented:
Hi @Hlavtox,

I have a case non treated following these steps:
1. Install ps_emailalert and set the threshold to 3
2. configure the Email 
3. Go to BO > Catalog > Create new product and enable it (by default, quantity = 0)
4. There is no email out of stock 

Without PR: the email is successfully sent after creating a new product out of stock.
![image](https://user-images.githubusercontent.com/16067358/135441056-3ce3d320-5649-49cb-a1eb-1c29e7544c67.png)

Thanks!

Hlavtox commented:
@khouloudbelguith Made a small mod, now it should send the email. But only if you have default activation state set to visible, because we don't send anything for disabled products.

Progi1984 commented on classes/stock/StockAvailable.php:
```diff
@@ -598,6 +598,7 @@ public static function setQuantity($id_product, $id_product_attribute, $quantity
                     $stockManager->saveMovement($id_product, $id_product_attribute, $deltaQuantity);
                 }
             } else {
+                $deltaQuantity = null;
```
```suggestion
```

Progi1984 commented on classes/stock/StockAvailable.php:
```diff
@@ -630,6 +631,7 @@ public static function setQuantity($id_product, $id_product_attribute, $quantity
                     'id_product' => $id_product,
                     'id_product_attribute' => $id_product_attribute,
                     'quantity' => $stock_available->quantity,
+                    'delta_quantity' => $deltaQuantity,
```
```suggestion
                    'delta_quantity' => $deltaQuantity ?? null,
```

Progi1984 reviewed (commented):
@Hlavtox Thank you for your contribution. I have brought you a small comment.

Hlavtox commented:
@khouloudbelguith Did you use the current version of the related PR?

Hlavtox commented:
@khouloudbelguith @Progi1984 I think we should discuss sending or not sending the emails in ps_emailalerts PR.
This PR provides modules with data and is working fine, so I think it should be merged...

khouloudbelguith commented:
> @khouloudbelguith Did you use the current version of the related PR?

Hi @Hlavtox,

Yes, I'm using this PR: https://github.com/PrestaShop/ps_emailalerts/pull/91 while testing this PR.

Thanks!

khouloudbelguith reviewed (dismissed):
Hi @Hlavtox,

In the `How to test` part, `Install both this and related PR`.
So I installed this PR using the PR module 
https://github.com/PrestaShop/ps_emailalerts/pull/91

I still have the same issue:
1. Go to BO > Catalog > Products
2. Create a product
3. Enable it
4. See error => there is no mail out of stock sent

I attached a screen

https://user-images.githubusercontent.com/16067358/152825536-19e4f2b7-2611-4107-b9d7-79a02421285c.mp4

**Without PR**

https://user-images.githubusercontent.com/16067358/152826153-320265f2-35b2-4979-a778-e0a9482aa758.mp4


As confirmed by the Product team, this issue should be fixed:
https://github.com/PrestaShop/PrestaShop/pull/25984#issuecomment-941004293

Thanks!

kpodemski commented:
Hello @khouloudbelguith @Hlavtox

I really don't think it requires more checks than having:

`PrestaShopLogger::addLog('Delta from '.date('d.m.Y H:i:s').': '.$params['delta']);`

inside `ps_qualityassurance` for the hook: `actionUpdateQuantity`.

This PR is about adding this information for the developer, and it's what it does. We can test PR related to `ps_emailalerts` later on.

I suggest the following:
- make sure that the delta information is correct using `ps_qualityassurance` module @khouloudbelguith 
- update the description and issue number to not involve `ps_emailalerts` module for now @Hlavtox 

What do you think?

Progi1984 commented:
@Hlavtox Could you update the `How to test?` for @PrestaShop/qa-functional team.

khouloudbelguith reviewed (commented):
H @Hlavtox, @kpodemski,

> I really don't think it requires more checks than having:
> 
> `PrestaShopLogger::addLog('Delta from '.date('d.m.Y H:i:s').': '.$params['delta_quantity']);`
> 
> inside `ps_qualityassurance` for the hook: `actionUpdateQuantity`.
> 
> This PR is about adding this information for the developer, and it's what it does. We can test PR related to `ps_emailalerts` later on.
> 
> I suggest the following:
> 
> * make sure that the `delta_quantity` information is correct using `ps_qualityassurance` module @khouloudbelguith
> * update the description and issue number to not involve `ps_emailalerts` module for now @Hlavtox
> 
> What do you think?

Thanks for your feedbacks.
I followed these steps:

1. Install this PR and Install the ps_qualityassurance module
2. Add the hook `actionUpdateQuantity` 
![image](https://user-images.githubusercontent.com/16067358/174277298-dbdf7dea-ee3a-410f-85bc-bb5b4fd8a4b2.png)
3. Create a new product or edit a product
4. In the BO > Advanced Parameters > Logs => The delta quantity is well displayed
![image](https://user-images.githubusercontent.com/16067358/174277845-5a3f9f14-dff1-479b-aa54-0b1c651c4e5d.png)

But as discussed with @Robin-Fischer-PS, thanks to update the `How to test `of this PR because the other [PR](https://github.com/PrestaShop/ps_emailalerts/pull/91) is NOK as described [here](https://github.com/PrestaShop/PrestaShop/pull/25984#pullrequestreview-874914197).
Also, thanks to provide us with more details about the possible impacts of this PR, so we can test and check all possible scenarios.

Thank you!

Hlavtox commented:
@khouloudbelguith Hello! PR description has been updated, we can finish that PR later, when this is merged. :-)

khouloudbelguith reviewed (commented):
Hi @Hlavtox,

Thanks for your update.
I found this issue

1. Install the ps_qualityassurance & add the hook `actionUpdateQuantity` ( `PrestaShopLogger::addLog('Delta from '.date('d.m.Y H:i:s').': '.$params['delta_quantity']);`)
2. Go to BO > Advanced Parameters > New & Experimental Features page && enable `New product page - Single store`
3. Go to Catalog > Products > Edit a product 
4. Change a quantity & save
5. Go to BO > Advanced Parmeters > Logs page 
6. check that no log added `Delta from 20.06.2022 16:46:11: 2`

https://user-images.githubusercontent.com/16067358/174630217-e712ee8a-1866-4fda-85fe-f531c8291a27.mp4

I tried with Product V1 => OK :heavy_check_mark: 
I tried to update from the Catalog > Stocks page => ok
I tried to update the quantity from the BO > Order view page => ok
I tried to refund an order => ok
I tried to create an order => ok

Thanks!

matks commented:
@Hlavtox friendly reminder 😄 QA team detected issues on this contribution

Hlavtox commented:
Hi @khouloudbelguith and @PrestaShop/prestashop-maintainers 

- I rebased this PR.
- Added a new hook call also to the new product page. It follows the same logic and is called on the same space as the old hook, just before `StockManager::saveMovement();`.
- I added id_shop to another hook call, to unify them.

kpodemski commented on src/Adapter/Product/Stock/Update/ProductStockUpdater.php:
```diff
@@ -309,5 +310,16 @@ private function saveMovement(StockAvailable $stockAvailable, StockModification
             ]
         );
+
+        Hook::exec(
```
Wow, that's quite a big issue if there was no hook previously

FabienPapet replied:
There should be dependency injection here.

kpodemski reviewed (commented):
@Hlavtox 

@FabienPapet is right. I didn't notice it's a Symfony service, you can safely add the next argument in the constructor and inject HookDispatcher 👍

Hlavtox committed:
Use dependency injection for hook manager

Hlavtox commented:
Hi @kpodemski @FabienPapet @marsaldev, I changed the implementation to DI as requested 👍
So if we can please reapprove and re-assign Waiting for QA.

jolelievre commented on src/Adapter/Product/Stock/Update/ProductStockUpdater.php:
```diff
@@ -309,5 +316,16 @@ private function saveMovement(StockAvailable $stockAvailable, StockModification
             ]
         );
+
+        $this->hookManager->exec(
```
Nice catch!! I didn't notice there was a missing hook from the static method, thanks

jolelievre reviewed (approved):
Thanks @Hlavtox

hibatallahAouadni reviewed (approved):
Hello @Hlavtox 

LGTM, QA :heavy_check_mark: 
![image](https://user-images.githubusercontent.com/23308427/202674386-60a1fa35-62fe-43c5-ac50-b6eafad8bee5.png)

Edit product qty from:

- Product V1 :heavy_check_mark: 
- Product V2 :heavy_check_mark: 
- Stock management :heavy_check_mark: 
- Create an Order (FO) :heavy_check_mark: 
- Create an Order (BO) :heavy_check_mark: 

Thanks!

kpodemski commented:
Thank you @Hlavtox @hibatallahAouadni :)
